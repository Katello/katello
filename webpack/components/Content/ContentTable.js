import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';

import { Table } from '../../move_to_foreman/components/common/table';
import { LoadingState } from '../../move_to_pf/LoadingState';

const ContentTable = ({ content, tableSchema, onPaginationChange }) => {
  const {
    loading, results, pagination, itemCount,
  } = content;

  const emptyStateData = {
    header: __('No Content found'),
  };

  return (
    <LoadingState
      loading={!results || loading}
      loadingText={__('Loading')}
    >
      <Table
        columns={tableSchema}
        rows={results}
        pagination={pagination}
        onPaginationChange={onPaginationChange}
        itemCount={itemCount}
        emptyState={emptyStateData}
      />
    </LoadingState>
  );
};

ContentTable.propTypes = {
  content: PropTypes.shape({
    loading: PropTypes.bool,
    // Disabling rule as existing code failed due to an eslint-plugin-react update
    // eslint-disable-next-line react/forbid-prop-types
    results: PropTypes.array,
    pagination: PropTypes.shape({}),
    itemCount: PropTypes.number,
  }).isRequired,
  onPaginationChange: PropTypes.func.isRequired,
  tableSchema: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};


export default ContentTable;
