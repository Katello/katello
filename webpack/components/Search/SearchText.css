.autocomplete-wrapper {
  width: 100%;
  position: relative;
  display: inline-block;
}

.autocomplete-wrapper > div {
  width: 100%;
}

.autocomplete-wrapper input {
  border: 1px solid #cecece;
  padding: 12px 15px;
  font-size: 18px;
  width: 100%;
}

.autocomplete-wrapper input:focus {
  border-color: #0F67FF;
  box-shadow: none;
  outline: none;
}

.autocomplete-wrapper .dropdown {
  width: 100%;
  padding: 0;
  text-align: left;
  max-height: 280px;
  overflow: hidden;
  overflow-y: auto;
  background-color: #ffffff;
  border: 1px solid #0F67FF;
  border-top: none;
}

.autocomplete-wrapper .item  {
  display: block;
  cursor: pointer;
  font-size: 18;
  padding: 5px;
}

.autocomplete-wrapper .item.selected-item {
  background-color: #0069ff;
  color: #FAFBFC;
}

.autocomplete-wrapper .item:hover {
  background-color: #0069ff;
  color: #FAFBFC;
}
