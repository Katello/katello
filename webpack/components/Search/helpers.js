// eslint-disable-next-line import/prefer-default-export
export const stringIncludes = (string, includes) => {
  const a = string.replace(/\s/g, '').toLowerCase();
  const b = includes.replace(/\s/g, '').toLowerCase();
  return a.includes(b);
};
