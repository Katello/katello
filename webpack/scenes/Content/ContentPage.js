import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { Grid, GridItem, TextContent, Text, TextVariants } from '@patternfly/react-core';
import { isEmpty } from 'lodash';
import { translate as __ } from 'foremanReact/common/I18n';
import { STATUS } from 'foremanReact/constants';
import ContentTable from './Table/ContentTable';
import { selectContentTypes, selectContentTypesStatus } from './ContentSelectors';
import ContentConfig from './ContentConfig';
import { getContentTypes } from './ContentActions';
import Loading from '../../components/Loading';
import EmptyStateMessage from '../../components/Table/EmptyStateMessage';

const ContentPage = () => {
  const dispatch = useDispatch();
  const contentTypesResponse = useSelector(selectContentTypes);
  const contentTypesStatus = useSelector(selectContentTypesStatus);
  const [selectedContentType, setSelectedContentType] = useState(null);
  const [showContentTypeSelector, setShowContentTypeSelector] = useState(true);
  const [contentTypes, setContentTypes] = useState(null);
  const { content_type: contentType } = useParams();

  useEffect(() => {
    // Set list of enabled content types and initial content type
    const buildContentTypes = () => {
      const types = {};
      contentTypesResponse.forEach((type) => {
        if (type.generic_browser) {
          const typeConfig = ContentConfig.find(config =>
            config.names.singularLabel === type.label);
          if (typeConfig) {
            const { names } = typeConfig;
            types[names.pluralTitle] = [names.singularLabel, names.pluralLabel];
          }
        }
      });
      return types;
    };

    if (contentTypesStatus === STATUS.RESOLVED) {
      const enabledContentTypes = buildContentTypes();
      if (!contentType) {
        setSelectedContentType(Object.keys(enabledContentTypes)[0]);
      } else {
        Object.entries(enabledContentTypes).forEach(([key, value]) => {
          if (value.includes(contentType)) {
            setSelectedContentType(key);
            setShowContentTypeSelector(false);
          }
        });
      }
      setContentTypes(enabledContentTypes);
    }
  }, [contentTypesStatus, contentTypesResponse, contentType]);

  useEffect(() => {
    dispatch(getContentTypes());
  }, [dispatch]);

  if (contentTypesStatus === STATUS.PENDING) {
    return <Loading />;
  } else if (isEmpty(contentTypes)) {
    return (
      <EmptyStateMessage
        title="Browsable content types not found."
        body="No content types are enabled for this page."
      />
    );
  }

  return (
    <Grid>
      <GridItem span={12} className="margin-24">
        <TextContent>
          <Text component={TextVariants.h1}>{__(`${selectedContentType}`)}</Text>
        </TextContent>
      </GridItem>
      <GridItem span={12}>
        <ContentTable {...{
          selectedContentType, setSelectedContentType, contentTypes, showContentTypeSelector,
        }}
        />
      </GridItem>
    </Grid>
  );
};

export default ContentPage;
