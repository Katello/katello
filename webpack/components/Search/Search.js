import React, { useState, useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { ControlLabel } from 'react-bootstrap';
import { loadSetting } from 'foremanReact/components/Settings/SettingsActions';
import { translate as __ } from 'foremanReact/common/I18n';
import PropTypes from 'prop-types';
import TypeAhead from '../TypeAhead';
import api, { foremanApi } from '../../services/api';
import { stringIncludes } from './helpers';
import {
  AUTOSEARCH_DELAY,
  AUTOSEARCH_WHILE_TYPING,
} from '../../scenes/Settings/SettingsConstants';

const Search = ({
  onSearch,
  updateSearchQuery,
  isDisabled,
  settings: { autoSearchDelay },
  initialInputValue,
  patternfly4,
  settings: { autoSearchEnabled },
  getAutoCompleteParams,
  foremanApiAutoComplete,
  bookmarkController,
}) => {
  const [items, setItems] = useState([]);
  const dispatch = useDispatch();
  const mountedRef = useRef(true);

  const onInputUpdate = async (searchTerm = '') => {
    const newItems = items.filter(({ text }) => stringIncludes(text, searchTerm));

    if (newItems.length !== items.length) {
      // Checking whether the current component is mounted before state change events
      if (!mountedRef.current) return;
      setItems(newItems);
    }

    const { params, endpoint } = getAutoCompleteParams(searchTerm);

    if (endpoint) {
      let data;
      if (foremanApiAutoComplete) {
        data = await foremanApi.get(endpoint, undefined, params);
      } else {
        data = await api.get(endpoint, undefined, params);
      }
      // Checking whether the current component is mounted before state change events
      if (!mountedRef.current) return;
      setItems(data?.data?.filter(({ error }) => !error).map(({ label }) => ({
        text: label.trim(),
      })));
    }

    if (autoSearchEnabled && patternfly4) {
      onSearch(searchTerm || '');
    }
  };

  useEffect(() => {
    dispatch(loadSetting(AUTOSEARCH_DELAY));
    dispatch(loadSetting(AUTOSEARCH_WHILE_TYPING));
    return () => { mountedRef.current = false; };
  }, [dispatch]);

  const onNewSearch = (search) => {
    if (updateSearchQuery) updateSearchQuery(search);
    onSearch(search);
  };

  return (
    <div>
      <ControlLabel srOnly>{__('Search')}</ControlLabel>
      <TypeAhead
        autoSearchDelay={autoSearchDelay}
        bookmarkController={bookmarkController}
        isDisabled={isDisabled}
        items={items}
        onInputUpdate={onInputUpdate}
        onSearch={onNewSearch}
        initialInputValue={initialInputValue}
        patternfly4={patternfly4}
        autoSearchEnabled={autoSearchEnabled}
      />
    </div>
  );
};

Search.propTypes = {
  onSearch: PropTypes.func.isRequired,
  getAutoCompleteParams: PropTypes.func.isRequired,
  foremanApiAutoComplete: PropTypes.bool,
  updateSearchQuery: PropTypes.func,
  initialInputValue: PropTypes.string,
  patternfly4: PropTypes.bool,
  isDisabled: PropTypes.bool,
  settings: PropTypes.shape({
    autoSearchEnabled: PropTypes.bool,
    autoSearchDelay: PropTypes.number,
  }),
  bookmarkController: PropTypes.string,
};

Search.defaultProps = {
  updateSearchQuery: undefined,
  foremanApiAutoComplete: false,
  initialInputValue: '',
  patternfly4: false,
  settings: {
    autoSearchEnabled: true,
  },
  isDisabled: undefined,
  bookmarkController: undefined,
};

export default Search;
