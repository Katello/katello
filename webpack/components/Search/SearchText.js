import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { SearchAutocomplete } from 'foremanReact/components/SearchBar/SearchAutocomplete';
import { useAPI } from 'foremanReact/common/hooks/API/APIHooks';
import { STATUS } from 'foremanReact/constants';
import { noop } from 'foremanReact/common/helpers';

const SearchText = ({
  data: {
    autocomplete: { url, apiParams } = { url: '' },
    disabled,
  },
  initialQuery,
  onSearchChange,
  name,
}) => {
  const [search, setSearch] = useState(initialQuery || '');
  const getAPIparams = input => ({ ...apiParams(input) });
  const { response, status, setAPIOptions } = useAPI('get', url, {
    params: getAPIparams(search),
  });
  const onChange = (newValue) => {
    onSearchChange(newValue);
    setSearch(newValue);
    setAPIOptions({ params: { ...getAPIparams(newValue) } });
  };
  const error =
    status === STATUS.ERROR || response?.[0]?.error
      ? response?.[0]?.error || response.message
      : null;

  let results = [];
  if (Array.isArray(response) && !error) {
    results = response.map(item => ({ label: item, category: '' }));
  }

  return (
    <div className="foreman-search-text">
      <SearchAutocomplete
        results={results}
        onSearchChange={onChange}
        value={search}
        disabled={disabled}
        error={error}
        name={name}
      />
    </div>
  );
};

SearchText.propTypes = {
  data: PropTypes.shape({
    autocomplete: PropTypes.shape({
      url: PropTypes.string.isRequired,
      apiParams: PropTypes.func,
    }).isRequired,
    disabled: PropTypes.bool,
  }).isRequired,
  initialQuery: PropTypes.string,
  onSearchChange: PropTypes.func,
  name: PropTypes.string,
};

SearchText.defaultProps = {
  initialQuery: '',
  onSearchChange: noop,
  name: null,
};

export default SearchText;
