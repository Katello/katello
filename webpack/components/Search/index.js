/* eslint-disable import/no-extraneous-dependencies */
import React, { Component } from 'react';
import { ControlLabel } from 'react-bootstrap';
import PropTypes from 'prop-types';

import TypeAhead from '../../move_to_pf/TypeAhead/TypeAhead';
import api from '../../services/api';
import { stringIncludes } from './helpers';

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = { items: [] };
    this.onInputUpdate = this.onInputUpdate.bind(this);
    this.onSearch = this.onSearch.bind(this);
  }

  componentDidMount() {
    this.onInputUpdate();
  }

  onInputUpdate(searchTerm = '') {
    const items = this.state.items.filter(({ text }) => stringIncludes(text, searchTerm));

    if (items.length !== this.state.items.length) {
      this.setState({ items });
    }

    const params = this.props.getAutoCompleteParams(searchTerm);
    const autoCompleteParams = [
      params.endpoint,
      params.headers || {},
      params.params || {},
    ];

    api.get(...autoCompleteParams).then(({ data }) => {
      this.setState({
        items: data.filter(({ error }) => !error).map(({ label }) => ({
          text: label.trim(),
        })),
      });
    });
  }

  onSearch(search) {
    if (this.props.updateSearchQuery) this.props.updateSearchQuery(search);
    this.props.onSearch(search);
  }

  render() {
    return (
      <div>
        <ControlLabel srOnly>{__('Search')}</ControlLabel>
        <TypeAhead
          items={this.state.items}
          onInputUpdate={this.onInputUpdate}
          onSearch={this.onSearch}
        />
      </div>
    );
  }
}

Search.propTypes = {
  /** Callback function when the "Search" button is pressed:
      onSearch(searchQuery)
   */
  onSearch: PropTypes.func.isRequired,
  /** Function returning params for the scoped-search complete api call:
      getAutoCompleteParams(searchQuery)

      Should return a shape { headers, params, endpoint }, e.g.:
      {
        headers: {},
        params: { organization_id, search },
        endpoint: '/subscriptions/auto_complete_search'
      }
  */
  getAutoCompleteParams: PropTypes.func.isRequired,
  updateSearchQuery: PropTypes.func,
};

Search.defaultProps = {
  updateSearchQuery: undefined,
};
export default Search;
