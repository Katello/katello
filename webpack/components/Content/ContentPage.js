import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Col, Row, Form, FormGroup } from 'react-bootstrap';
import Search from '../../components/Search/index';
import ContentTable from './ContentTable';

const ContentPage = ({
  header, onSearch, getAutoCompleteParams,
  updateSearchQuery, initialInputValue,
  content, tableSchema, onPaginationChange,
}) => (
  <Grid bsClass="container-fluid">
    <Row>
      <Col sm={12}>
        <h1>{header}</h1>
      </Col>
    </Row>
    <Row>
      <Col sm={6}>
        <Form className="toolbar-pf-actions">
          <FormGroup className="toolbar-pf toolbar-pf-filter">
            <Search
              onSearch={onSearch}
              getAutoCompleteParams={getAutoCompleteParams}
              updateSearchQuery={updateSearchQuery}
              initialInputValue={initialInputValue}
            />
          </FormGroup>
        </Form>
      </Col>
    </Row>
    <Row>
      <Col sm={12}>
        <ContentTable
          content={content}
          tableSchema={tableSchema}
          onPaginationChange={onPaginationChange}
        />
      </Col>
    </Row>
  </Grid>
);

ContentPage.propTypes = {
  header: PropTypes.string.isRequired,
  content: PropTypes.shape({}).isRequired,
  tableSchema: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  onSearch: PropTypes.func.isRequired,
  getAutoCompleteParams: PropTypes.func.isRequired,
  updateSearchQuery: PropTypes.func.isRequired,
  initialInputValue: PropTypes.string.isRequired,
  onPaginationChange: PropTypes.func.isRequired,
};

export default ContentPage;
