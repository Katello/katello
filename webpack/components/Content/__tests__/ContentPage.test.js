import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ContentPage from '../ContentPage';
import ContentTable from '../ContentTable';
import Search from '../../../components/Search/index';

jest.mock('foremanReact/components/Pagination/PaginationWrapper', () => (<div>Pagination Mock</div>));

describe('Content page', () => {
  it('should render and contain appropriate components', async () => {
    const contentHeader = 'Content Header';
    const content = {};
    const onSearch = () => {};
    const getAutoCompleteParams = () => {};
    const updateSearchQuery = () => {};
    const searchQuery = '';
    const onPaginationChange = () => {};
    const TableSchema = [];

    const wrapper = shallow(<ContentPage
      header={contentHeader}
      content={content}
      tableSchema={TableSchema}
      onSearch={onSearch}
      getAutoCompleteParams={getAutoCompleteParams}
      updateSearchQuery={updateSearchQuery}
      initialInputValue={searchQuery}
      onPaginationChange={onPaginationChange}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(ContentTable)).toHaveLength(1);
    expect(wrapper.find(Search)).toHaveLength(1);
  });
});
