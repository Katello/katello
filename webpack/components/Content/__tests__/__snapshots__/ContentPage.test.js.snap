// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content page should render and contain appropriate components 1`] = `
<Grid
  bsClass="container-fluid"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={12}
    >
      <h1>
        Content Header
      </h1>
    </Col>
  </Row>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={6}
    >
      <Form
        bsClass="form"
        className="toolbar-pf-actions"
        componentClass="form"
        horizontal={false}
        inline={false}
      >
        <FormGroup
          bsClass="form-group"
          className="toolbar-pf toolbar-pf-filter"
        >
          <Search
            getAutoCompleteParams={[Function]}
            initialInputValue=""
            onSearch={[Function]}
            updateSearchQuery={[Function]}
          />
        </FormGroup>
      </Form>
    </Col>
  </Row>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={12}
    >
      <ContentTable
        content={Object {}}
        onPaginationChange={[Function]}
        tableSchema={Array []}
      />
    </Col>
  </Row>
</Grid>
`;
